### PR TITLE
Extract Debian license data from copyright text

### PR DESCRIPTION
**Pull request by mukultaneja**
This PR does following,

1. It proposed a license extraction logic based
on `pkg_format`.
2. It proposed a change in `base.yml` file to extract
copyright file location instead of copyright text.
3. It adds `pkg_licenses` property in `Package`
class to record multiple liceses in a package.
4. It updates default report generator to
print multiple licenses of a packge.
5. It updates default generator to address
'to many brances' build error.

Work towards: #499

Signed-off-by: mukultaneja <mtaneja@vmware.com>

### CODE REVIEW

File: tern/command_lib/base.yml [hunk 1]
```diff
@@ -94,7 +94,7 @@ dpkg:
       1:
         container:
           - "pkgs=`dpkg --get-selections | cut -f1 -d':' | awk '{print $1}'`"
-          - "for p in $pkgs; do cat /usr/share/doc/$p/copyright; echo LICF; done"
+          - "for p in $pkgs; do /bin/cat /usr/share/doc/$p/copyright; echo LICF; done"
```
Comment (nishakm) [inline tern/command_lib/base.yml:97]: Is this change more related to resolving #565 ?

Comment (rnjudge) [reply tern/command_lib/base.yml:97]: If so, please move to its own PR :)

Comment (mukultaneja) [reply tern/command_lib/base.yml:97]: yes, it will address issue #565 as well. Here, I have submitted this change because I am using `debut` external utility to parse Debian copyright data which accepts copyright file location instead of copyright data.

Comment (rnjudge) [reply tern/command_lib/base.yml:97]: Not sure what @nishakm thinks but I personally think that if it's addressing another bug it should be it's own PR or at least a separate commit in this PR. Your commit message says "It proposed a change in base.yml file to extract copyright file location instead of copyright text." but you also say that this is addressing but #565. The change in this PR only seems to be changing/specifying the location of the cat utility (which addresses #565) -- it's not clear to me how this changes the information that is retrieved as your commit message suggests.

Comment (mukultaneja) [reply tern/command_lib/base.yml:97]: @rnjudge I can understand your point is but the two changes which I proposed through this PR are related to each other. If I don't make a change in `base.yml`, I can't parse copyright text of any Debian package using `debut` utility. It is a pure coincidence that another issue is already being notified which requires the same change of `base.yml`. Although, I am ok to dedicate this `base.yml` change to issue #565. Anyway, the change in `base.yml` is the base of parsing copyright text.

Comment (rnjudge) [reply tern/command_lib/base.yml:97]: If you want to keep the commits together that's fine but can you update your commit message to more accurately describe the changes you're making? The change to base.yml doesn't `extract copyright file location instead of copyright text` itself -- it may enable debut to do this -- but the change in base.yml as far as Tern code base is concerned is only specifying the location of the cat utility. Please also mention that this simultaneously works towards issue #565 somewhere in your commit message. Thank you!

Comment (mukultaneja) [reply tern/command_lib/base.yml:97]: Sure!

Comment (rnjudge) [reply tern/command_lib/base.yml:97]: Thank you! Really appreciate this PR!

File: tern/analyze/common.py [hunk 2]
```diff
@@ -249,6 +250,24 @@ def add_base_packages(image_layer, binary, shell):
...
         # get all the packages in the base layer
         pkg_dict, invoke_msg, warnings = collate_list_metadata(shell, listing)
+
+        if listing.get("pkg_format") == "deb":
+            pkg_dict["licenses"] = list()
+            pd = os.path.join(
+                rootfs.mount_dir, constants.temp_folder, constants.mergedir)
+            for copyright in pkg_dict["copyrights"]:
+                cp = '{0}{1}'.format(pd, copyright.strip())
+                d = DebianCopyright.from_file(cp)
```
Comment (nishakm) [inline tern/analyze/common.py:260]: I wonder if DebianCopyright has a parser that would read data from `pkg_dict["copyright"]` instead. If reading it from file is more accurate, we may want to create a separate python script for it.

Comment (mukultaneja) [reply tern/analyze/common.py:260]: I think it is a good idea to abstract this function call using a custom module/script. It may also help in a situation where we may want to add more filtration for license text in the future.

Comment (nishakm) [reply tern/analyze/common.py:260]: That is the rub. Typically, a package has a declared license (see rpm, pip, etc). In Debian's case, their licenses are listed in the copyright text of the package. So this is really post processing of the licenses in a Debian specific copyright file. This is why I asked if `debut` can parse any individual copyright text and get a list of licenses for it. If it can, you can just write a function to do it and call it here.

Comment (mukultaneja) [reply tern/analyze/common.py:260]: @nishakm how to handle `Unable to import 'debut.copyright'` lint error here?

Comment (nishakm) [reply tern/analyze/common.py:260]: you would have to install tern first in your virtual environment `pip install -e.[dev]`.

Comment (mukultaneja) [reply tern/analyze/common.py:260]: I am sorry but I was referring circleci linting errors.

Comment (nishakm) [reply tern/analyze/common.py:260]: You will have to add `debut` to `requirements.txt` and `requirements.in`.

File: tern/command_lib/base.yml (hunk 1 above)
Comment (mukultaneja) [reply tern/command_lib/base.yml]: @rnjudge, Done! I have moved the `base.yml` to a separate pull request.

File: tern/analyze/common.py (hunk 2 above)
Comment (mukultaneja) [reply tern/analyze/common.py]: @nishakm, I have updated the code with requested changes. Please take a look and share your feedback.

File: requirements.txt [hunk 3]
```diff
@@ -12,3 +12,4 @@ dockerfile-parse~=0.0
 requests~=2.23
 stevedore>=1.32
 pbr>=5.4
+debut
```
Comment (nishakm) [inline requirements.txt:15]: Can you give me the version of debut you were using at the point you created the PR and a list of dependencies?

File: tern/analyze/common.py [hunk 4]
```diff
@@ -222,6 +223,59 @@ def convert_to_pkg_dicts(pkg_dict):
...
+    '''
+    Given the debian copyright file location,
+    1. parse the copyright file
+    2. filter out all the available licenses
+    3. returns a list of unique licenses found inside
+    the copyright text
+    '''
+    mounted_path = os.path.join(
+        rootfs.mount_dir, constants.temp_folder, constants.mergedir)
+    deb_copyright = '{0}{1}'.format(mounted_path, deb_copyright)
```
Comment (nishakm) [inline tern/analyze/common.py:236]: This is interesting: what gets returned by `collate_list_metadata` should be a dictionary with a key called `copyrights` and these should be copyright text not copyright file locations. So what `deb_copyright` here is a string that looks like `.tern/temp/mergedir<some text>`. But `DebianCopyright.from_file` requires a file location only.

File: tern/analyze/common.py [hunk 5]
```diff
@@ -222,6 +223,59 @@ def convert_to_pkg_dicts(pkg_dict):
...
+    deb_pkg_data = DebianCopyright.from_file(deb_copyright).to_dict()
```
Comment (nishakm) [inline tern/analyze/common.py:237]: From the above inference, I don't think we can use this. And it looks like debut doesn't have a direct way of instantiating this class given just a debian copyright text. However, there also seems to be `debcon.get_paragraphs_data` [here](https://github.com/nexB/debut/blob/972761071d70a2665a434c9df83fdaf371c338ee/src/debut/debcon.py#L502) which we can possibly use directly. It might be worthwhile to look into this first.

Comment (mukultaneja) [reply tern/analyze/common.py:237]: @nishakm the link which you pointed out to `get_paragraph_data_from_file` was my earlier proposal in the notebook but you suggested going in this way.

File: tern/analyze/common.py (hunk 4 above)
Comment (mukultaneja) [reply tern/analyze/common.py:236]: @nishakm if we see this code in the combination of updated `base.yml` with returning copyright file location then it would make complete sense. I did this because as you pointed out that `debut` does not have a way as of now to use the copyright text directly.

File: requirements.txt (hunk 3 above)
Comment (mukultaneja) [reply requirements.txt:15]: @nishakm, here along with `debut` version do I need to add its dependencies also in `requirements.txt`?

File: tern/analyze/common.py (hunk 5 above)
Comment (nishakm) [reply tern/analyze/common.py:237]: Yes, `get_paragraph_data_from_file` reads a file and then calls `get_paragraphs_data` on the read text. This is the functionality I would rather use, instead of referring to the file path. This helps with our separation of how individual OSs choose to record this metadata. I haven't tested out `get_paragraphs_data` but you could try it out and see if it works.

File: requirements.txt (hunk 3 above)
Comment (nishakm) [reply requirements.txt]: Nope. We just need to figure out how much the dependencies overlap with the packages we are already using. If there is no overlap then we can add `debut>=major.minor` where major.minor is the version you have. If there is overlap, we will have to add it as `debut~=major.minor`.

File: tern/analyze/common.py (hunk 4 above)
Comment (nishakm) [reply tern/analyze/common.py:236]: I would rather not have that change as I would like the enabling on `base.yml` to be as uniform across all package managers as possible.

File: requirements.txt (hunk 3 above)
Comment (mukultaneja) [reply requirements.txt]: @nishakm I have updated the `debut` version in `requirements.txt`. Please take a look once

File: requirements.txt [hunk 6]
```diff
@@ -12,3 +12,4 @@ dockerfile-parse~=0.0
 requests~=2.23
 stevedore>=1.32
 pbr>=5.4
+debut==0.9.4
```
Comment (nishakm) [inline requirements.txt:15]: Are any of debut's dependencies overlapping with tern's? If no then please change this to `debut>=0.9` and if yes then please change it to `debut~=0.9`. Thanks!

Comment (mukultaneja) [reply requirements.txt]: Done!

File: tern/analyze/common.py (hunk 5 above)
Comment (mukultaneja) [reply tern/analyze/common.py]: Done! Please review the update file.

File: tern/analyze/common.py [hunk 7]
```diff
@@ -222,6 +228,68 @@ def convert_to_pkg_dicts(pkg_dict):
...
+    '''
+    collected_paragraphs = list()
+    pkg_licenses = set()
+    for paragraph in iter(debcon.get_paragraphs_data(deb_copyright)):
+        if 'format' in paragraph or 'format-specification' in paragraph:
+            cp = CopyrightHeaderParagraph.from_dict(paragraph)
+        elif 'files' in paragraph:
+            cp = CopyrightFilesParagraph.from_dict(paragraph)
+        elif 'license' in paragraph:
+            cp = CopyrightLicenseParagraph.from_dict(paragraph)
```
Comment (nishakm) [inline tern/analyze/common.py:247]: I wonder if we can just use this.

File: tern/analyze/common.py [hunk 8]
```diff
@@ -222,6 +228,68 @@ def convert_to_pkg_dicts(pkg_dict):
...
+        else:
+            cp = CatchAllParagraph.from_dict(paragraph)
+        collected_paragraphs.append(cp)
+
+    deb_pkg_data = DebianCopyright(collected_paragraphs).to_dict()
+    for paragraph in deb_pkg_data.get("paragraphs"):
+        pkg_license = paragraph.get("license") \
+            if 'license' in paragraph else ''
```
Comment (nishakm) [inline tern/analyze/common.py:255]: Because we seem to be doing the same check here, regardless of which paragraphs we get.

File: tern/analyze/common.py [hunk 9]
```diff
@@ -249,6 +317,11 @@ def add_base_packages(image_layer, binary, shell):
             shell = constants.shell
         # get all the packages in the base layer
         pkg_dict, invoke_msg, warnings = collate_list_metadata(shell, listing)
+
+        if listing.get("pkg_format") == "deb":
+            pkg_dict["licenses"] = get_deb_package_licenses(
```
Comment (nishakm) [inline tern/analyze/common.py:322]: Here's where our logic fails: we're assuming that each package has 1 declared license, but for Debian, one package can have many licenses. We would like to associate the licenses to the package it came from. How do we do that if we can only store 1 license string in the `Package` class?

Comment (nishakm) [reply tern/analyze/common.py:322]: I would suggest to create a new property in `Package` called "licenses" to store multiple licenses if they can be found.

File: requirements.txt [hunk 10]
```diff
@@ -12,3 +12,4 @@ dockerfile-parse~=0.0
 requests~=2.23
 stevedore>=1.32
 pbr>=5.4
+debut>=0.9.4
```
[the remark this answers is not in the document]
Comment (mukultaneja) [reply requirements.txt]: Done!

File: tern/analyze/common.py [hunk 11]
```diff
@@ -23,6 +23,12 @@
 from tern.utils import constants
 from tern.utils import general
 from tern.utils import rootfs
+from debut import debcon
+from debut.copyright import DebianCopyright
+from debut.copyright import CatchAllParagraph
+from debut.copyright import CopyrightHeaderParagraph
+from debut.copyright import CopyrightFilesParagraph
+from debut.copyright import CopyrightLicenseParagraph
```
[the remark this answers is not in the document]
Comment (mukultaneja) [reply tern/analyze/common.py]: Done! I am using `from debut import copyright as debut_copyright` because of using built-in copyright build error.

File: tern/analyze/common.py [hunk 12]
```diff
@@ -222,6 +228,68 @@ def convert_to_pkg_dicts(pkg_dict):
     return pkg_list
 
 
+def get_deb_package_license(deb_copyright):
```
[the remark this answers is not in the document]
Comment (mukultaneja) [reply tern/analyze/common.py]: I have updated the function name with `get_licenses_from_deb_copyright`

File: tern/analyze/common.py (hunk 7 above)
Comment (mukultaneja) [reply tern/analyze/common.py]: Done! I have removed every other parsing function call except `CopyrightLicenseParagraph`.

File: tern/analyze/common.py (hunk 8 above)
Comment (mukultaneja) [reply tern/analyze/common.py]: Done!

File: tern/analyze/common.py (hunk 9 above)
Comment (mukultaneja) [reply tern/analyze/common.py]: > Here's where our logic fails: we're assuming that each package has 1 declared license, but for Debian, one package can have many licenses. We would like to associate the licenses to the package it came from. How do we do that if we can only store 1 license string in the `Package` class?

I can get your point here that is why the previous commits in this PR collect all the licenses of a package and assign `pkg_license` property a string of all licenses collected. So say if a package `dpkg` got two licenses `['GPL-V2, LGPL V3+']` then the package's  `pkg_license` property would be having a string value `GPL-V2, LGPL V3+` which means a package would be associated with its respective licenses.

Comment (mukultaneja) [reply tern/analyze/common.py]: > I would suggest to create a new property in `Package` called "licenses" to store multiple licenses if they can be found.

I have incorporated this change in the latest commit. Now I am having `pkg_licenses` property in the `Package` class and storing package licenses inside it. This change is also required to update the default report generator to print `pkg_licenses` if we have any so made that change also in the latest commit.

**Comment by nishakm** [comment]
@mukultaneja There might be some minor cleaning up needed. Would you mind if I submit this commit for you with the clean-up changes on top of it?

**Comment by mukultaneja** [comment]
> @mukultaneja There might be some minor cleaning up needed. Would you mind if I submit this commit for you with the clean-up changes on top of it?

@nishakm, thanks for reviewing the changes. If it's easy to define the cleanup changes I would like to take them and finish the PR but if they are hard to define then I guess you can better take the decision :)

**Comment by rnjudge** [comment]
Thanks for all your work on this @mukultaneja. Closing this PR as https://github.com/tern-tools/tern/pull/694 replaces it.
